src/output-tx.c

```diff
@@ -138,35 +148,87 @@ static TmEcode OutputTxLog(ThreadVars *tv, Packet *p, void *thread_data, PacketQ
         if (!(AppLayerParserStateIssetFlag(f->alparser, APP_LAYER_PARSER_EOF)))
         {
             int tx_progress = AppLayerParserGetStateProgress(p->proto, alproto,
-                                                             tx, FlowGetDisruptionFlags(f, STREAM_TOSERVER));
-            if (tx_progress < tx_progress_done_value_ts) {
-                SCLogDebug("progress not far enough, not logging");
-                break;
+                tx, FlowGetDisruptionFlags(f, STREAM_TOSERVER));
+            if (tx_progress == tx_progress_done_value_ts) {
+                ts_ready = 1;
             }
 
             tx_progress = AppLayerParserGetStateProgress(p->proto, alproto,
-                                                         tx, FlowGetDisruptionFlags(f, STREAM_TOCLIENT));
-            if (tx_progress < tx_progress_done_value_tc) {
-                SCLogDebug("progress not far enough, not logging");
-                break;
+                tx, FlowGetDisruptionFlags(f, STREAM_TOCLIENT));
+            if (tx_progress == tx_progress_done_value_tc) {
+                tc_ready = 1;
             }
         }
 
+        if (!(ts_ready || tc_ready)) {
+            SCLogNotice("progress not for enough, not logging");
+            break;
+        }
+
+        /* Prevent the processing of one direction getting farther
```

Comment:
On 17-12-15 16:42, Jason Ish wrote:
> In src/output-tx.c
> <https://github.com/inliniac/suricata/pull/1794#discussion_r47920316>:
> 
>>              }
>>          }
>>  
>> +        if (!(ts_ready || tc_ready)) {
>> +            SCLogNotice("progress not for enough, not logging");
>> +            break;
>> +        }
>> +
>> +        /* Prevent the processing of one direction getting farther
> 
> All this does is wait for one transaction to be complete before moving
> on to the other. I thought that was current behaviour. For example,
> transaction can go like this:
> 
>  1. TX 0 Request
>  2. TX 1 Request
>  3. TX 1 Response
>  4. TX 0 Response
> 
> Now that we check per direction, without enforcing order, the response
> for tx 0 will never get logged as the transaction ID tracker is
> incremented after step 3, so we never check the state of of TX 0 again.
> 
> I believe this was implicit before, now that we check per direction we
> have to enforce it explicitly to avoid the transaction counters being
> incremented to soon.
> 
> Regarding HTTP - is that something that happens now? Or is that
> something we could do with per direction logging. I think that will also
> require a different way of tracking complete transactions like my
> overlapping TX case.

In HTTP the transactions are implicit, so the first response is mapped
to the first request. So we can't have a case there the response for a
TX 1 comes in before TX 0.